stable_baselines3/ppo/ppo.py

```diff
@@ -231,18 +233,28 @@ def train(self) -> None:
 
                 loss = policy_loss + self.ent_coef * entropy_loss + self.vf_coef * value_loss
 
+                # Calculate approximate form of reverse KL Divergence for early stopping
+                # see issue #417: https://github.com/DLR-RM/stable-baselines3/issues/417
+                # and Schulman blog: http://joschu.net/blog/kl-approx.html
+                with th.no_grad():
+                    r = rollout_data.old_log_prob - log_prob
```

Comment:
maybe a better variable name would be `prob_ratio` or just `ratio` (but this name is used before)

Reply:
Sure, I think `log_ratio` might be more appropriate?

Reply:
yep =)

Reply:
Should this not also be the other way around `log_ratio = log_prob - rollout_data.old_log_prob`? In the material `r = log p(x)/q(x)`, where data is sampled from `q(x)`, which here corresponds to the old log probs.

Reply:
I don't think so, let me elaborate below:

`p(x)` represents a 'true' distribution and `q(x)` is the distribution estimating `p(x)`. You're right, we sample data from `q(x)`, in this case that means sampled from the neural network with respect to the network parameters, θ. That is, `q(x) = current_dist`, which is estimating the old policy such that `p(x) = old_dist`

Reply:
Hmm I am having hard time wrapping my head around that. If we sample from `q`, it should then be `q(x) = old_dist`, because it was that `old_dist` we sampled the experiences from. Also looking at the replaced KL estimate, which is `log q(x)/p(x)` (or `rollout_data.old_log_prob - log_prob`), seems to indicate we should treat `old_log_prob` as `log q(x)`.

Reply:
From the article I linked in the issue on Forward and Reverse KL: 
![image](https://user-images.githubusercontent.com/47002898/117313009-6bf07d80-ae7d-11eb-8489-821027eea77d.png)

In this case, I'd argue we control the current policy through the network parameters. I think the 'true distribution' is the old policy since this is the fixed benchmark we want to measure the current policy against.

Reply:
Another little blurb:
![image](https://user-images.githubusercontent.com/47002898/117314117-5c256900-ae7e-11eb-95bc-637747fc9646.png)

Reply:
Although I see your point, the neural network is used to predict both the old and new policy, so changing the network parameters would in fact change the old policy as well, but we have to take one of these as fixed, and I think it makes more sense to take the old policy for this role.

Reply:
Gah I am still having trouble digesting all this (been a long time since I had to play with KLs in this way) and still not convinced with the setup. My main problem is with how Schulman's blog sets `x ~ q(x)`, which in our case is the old policy (which is also fixed with stored `log_prob`), but the code now treats as samples are from `x ~ p(x)`, changing things around.

Could you do a quick comparison what happens in the CartPole KL-plots if you flip the ratio around? Maybe this could shed bit of light in to the matter.

@araffin can you throw a comment here?

Reply:
Yes actually I do see your point, the way sampling is being described certainly complicates things.

However, I think a good thought experiment is to zoom out briefly and discuss the KL divergence without the approximation, since this setting removes the idea of samples altogether 😁:

`KL(x) = E[log(q(x) / p(x))]` with the expectation over samples from q, so `KL(x) = sum/integral_x q(x) * q(x)/p(x)`. 

I'd argue the way to calculate this in this setting would be to get all the possible actions, `x`, and put them in that equation where `q(x)` is the current policy and `p(x)` is the old policy, again since we control the current policy. If this is the case for the full form equation, I'd propose it should also be the case for the approximation.

Still, I tried reversing the ratio to see what happens but as expected the ratio has the same range, and you get a similar looking graph:
![image](https://user-images.githubusercontent.com/47002898/117426728-2636c200-af1c-11eb-9a8e-b52266be1ed4.png)

Reply:
Thanks for being so patient with slow-me! I do agree with your explanation and that old policy should be `p(x)`, so lets go with this current code you have. The empirical results are still desirable, with a major change in value scale. While I do not normally agree with "results look good enough" approach, I think we need to call this done for now before we get too stuck into this (and I do not have the skill-set to fully argue against this).

Lets leave this discussion open ("not resolved") so anybody wondering this change can see the discussion :).

Reply:
Not at all to be honest I'm less sure now myself 😅, I agree with the way sampling is phrased it does appear that it should be the other way round; however, I still can't shake the feeling that it breaks intuition (for me at least). 

I agree, let's leave it open. Hopefully Schulman one day stumbles onto this and maybe sheds some light on the issue lool.

Reply:
I'll the brave move and poke @joschu if he would be so kind to help us lost souls on this matter ^^

Reply:
Glad to see the blog post is useful! Using the blog's notation, I'd let q be the old distribution (=the sampling distribution). Then we can approximate KL[new,old] or KL[old,new] by using the two expressions I listed.
I assume you guys know this, so maybe I'm missing something.

Reply:
Ah ok, I thought it was the other way round (q = new/current distribution) because I think it fits the intuition of KL Divergence better (we control the current distribution and the old distribution is seen as fixed). 

Just one more question if you'd be willing, in the case of the full KL Divergence (i.e. away from sampling since we sum/integrate over all actions), would you still say q = old distribution? If so, why?